pkg/platform/kube/provisioner/apigateway/provisioner.go

```diff
@@ -0,0 +1,369 @@
+		var baseUpstream platform.APIGatewayUpstreamSpec
+
+		// determine which upstream is the canary one
+		if upstreams[0].Percentage != 0 {
+			baseUpstream = upstreams[1]
+			canaryUpstream = upstreams[0]
+		} else if upstreams[1].Percentage != 0 {
+			baseUpstream = upstreams[0]
+			canaryUpstream = upstreams[1]
+		} else {
+			return errors.New("Percentage must be set on one of the upstreams (canary)")
+		}
+
+		// validity check
+		if canaryUpstream.Percentage > 100 || canaryUpstream.Percentage < 1 {
+			return errors.New("The canary upstream percentage must be between 1 and 100")
+		}
+
+		// add the base ingress
+		baseIngressResources, err := p.generateNginxIngress(ctx, apiGateway, baseUpstream)
+		if err != nil {
+			return errors.Wrap(err, "Failed to generate the base nginx ingress")
+		}
+		ingresses[baseIngressResources.Ingress.Name] = baseIngressResources
+
+		// add the canary ingress
+		canaryIngressResources, err := p.generateNginxIngress(ctx, apiGateway, canaryUpstream)
+		if err != nil {
+			return errors.Wrap(err, "Failed to generate the canary nginx ingress")
+		}
+		ingresses[canaryIngressResources.Ingress.Name] = canaryIngressResources
+	}
+
+	// create ingresses
+	for ingressName, ingressResources := range ingresses {
+		if _, _, err := p.ingressManager.CreateOrUpdateIngressResources(ingressResources); err != nil {
+			p.Logger.WarnWithCtx(ctx, "Failed to create/update api-gateway ingress resources",
+				"err", errors.Cause(err),
+				"ingressName", ingressName,
+				"appliedIngressNames", appliedIngressNames)
+			return errors.New("Failed to create/update api-gateway ingress resources")
+		}
+
+		appliedIngressNames = append(appliedIngressNames, ingressName)
+	}
+
+	// sleep 4 seconds as a safety, so nginx will finish updating the ingresses properly (it takes time)
+	p.Logger.DebugWithCtx(ctx,
+		"Updated nginx ingresses, sleeping for 4 seconds so nginx will stabilize",
+		"apiGatewayName", apiGateway.Name)
+	time.Sleep(4 * time.Second)
+
+	return nil
+}
+
+func (p *Provisioner) DeleteAPIGateway(ctx context.Context, namespace, name string) {
+
+	p.Logger.DebugWithCtx(ctx,
+		"Deleting api-gateway base ingress",
+		"name", name)
+
+	err := p.ingressManager.DeleteIngressByName(p.generateIngressName(name, false), namespace, true)
+	if err != nil {
+		p.Logger.WarnWithCtx(ctx, "Failed to delete base ingress. Continuing with deletion",
+			"err", errors.Cause(err))
+	}
+
+	p.Logger.DebugWithCtx(ctx,
+		"Deleting api-gateway canary ingress",
+		"name", name)
+
+	err = p.ingressManager.DeleteIngressByName(p.generateIngressName(name, true), namespace, true)
+	if err != nil {
+		p.Logger.WarnWithCtx(ctx, "Failed to delete canary ingress. Continuing with deletion",
+			"err", errors.Cause(err))
+	}
+}
+
+func (p *Provisioner) tryRemovePreviousCanaryIngress(ctx context.Context, apiGateway *nuclioio.NuclioAPIGateway) {
+	p.Logger.DebugWithCtx(ctx,
+		"Trying to remove previous canary ingress",
```

Comment:
please move log lines to first and rest of var/varnames to newlines if line buffer is exceeded